src/dbnode/encoding/proto/encoder.go

```diff
@@ -0,0 +1,910 @@
+				return err
+			}
+		case isCustomIntEncodedField(customField.fieldType):
+			if err := enc.encodeIntValue(i, iVal); err != nil {
+				return err
+			}
+		case customField.fieldType == cBytes:
+			if err := enc.encodeBytesValue(i, iVal); err != nil {
+				return err
+			}
+		default:
+			customEncoded = false
+		}
+
+		if customEncoded {
+			// Remove the field from the message so we don't include it
+			// in the proto marshal.
+			m.ClearFieldByNumber(customField.fieldNum)
+		}
+	}
+	enc.hasEncodedFirstSetOfCustomValues = true
+
+	return nil
+}
+
+func (enc *Encoder) encodeTSZValue(i int, iVal interface{}) error {
+	var (
+		val         float64
+		customField = enc.customFields[i]
+	)
+	switch typedVal := iVal.(type) {
+	case float64:
+		val = typedVal
+	case float32:
+		val = float64(typedVal)
+	default:
+		return fmt.Errorf(
+			"%s found unknown type in fieldNum %d", encErrPrefix, customField.fieldNum)
+	}
+
+	enc.customFields[i].floatXORState.WriteFloat(enc.stream, val)
+	return nil
+}
+
+func (enc *Encoder) encodeIntValue(i int, iVal interface{}) error {
+	var (
+		signedVal   int64
+		unsignedVal uint64
+		customField = enc.customFields[i]
+	)
+	switch typedVal := iVal.(type) {
+	case uint64:
+		unsignedVal = typedVal
+	case uint32:
+		unsignedVal = uint64(typedVal)
+	case int64:
+		signedVal = typedVal
+	case int32:
+		signedVal = int64(typedVal)
+	default:
+		return fmt.Errorf(
+			"%s found unknown type in fieldNum %d", encErrPrefix, customField.fieldNum)
+	}
+
+	if isUnsignedInt(customField.fieldType) {
+		if !enc.hasEncodedFirstSetOfCustomValues {
+			enc.encodeFirstUnsignedIntValue(i, unsignedVal)
+		} else {
+			enc.encodeNextUnsignedIntValue(i, unsignedVal)
+		}
+	} else {
+		if !enc.hasEncodedFirstSetOfCustomValues {
+			enc.encodeFirstSignedIntValue(i, signedVal)
+		} else {
+			enc.encodeNextSignedIntValue(i, signedVal)
+		}
+	}
+
+	return nil
+}
+
+func (enc *Encoder) encodeBytesValue(i int, iVal interface{}) error {
+	customField := enc.customFields[i]
+	currBytes, ok := iVal.([]byte)
+	if !ok {
+		currString, ok := iVal.(string)
+		if !ok {
+			return fmt.Errorf(
+				"%s found unknown type in fieldNum %d", encErrPrefix, customField.fieldNum)
+		}
+		currBytes = []byte(currString)
+	}
+
+	var (
+		hash             = xxhash.Sum64(currBytes)
+		numPreviousBytes = len(customField.bytesFieldDict)
+		lastStateIdx     = numPreviousBytes - 1
+		lastState        encoderBytesFieldDictState
+	)
+	if numPreviousBytes > 0 {
+		lastState = customField.bytesFieldDict[lastStateIdx]
+	}
+
+	if numPreviousBytes > 0 && hash == lastState.hash {
+		streamBytes, _ := enc.stream.Rawbytes()
+		match, err := enc.bytesMatchEncodedDictionaryValue(
+			streamBytes, lastState, currBytes)
+		if err != nil {
+			return fmt.Errorf(
+				"%s error checking if bytes match last encoded dictionary bytes: %v",
+				encErrPrefix, err)
+		}
+		if match {
+			// No changes control bit.
+			enc.stream.WriteBit(opCodeNoChange)
```

Comment:
Do we have some tests to validate that matches are occurring as expected etc?

Reply:
Not right now, I'm gonna add a Stats() method to the encoder so we can keep track of how well the compression is working in a follow up P.R and I'll use that to update the tests.

Tracking here: https://github.com/m3db/m3/issues/1471

Reply:
Sweet, sounds good. Yeah some visibility, either in test or prod would be awesome.